tsl/src/continuous_aggs/invalidation.c

```diff
@@ -0,0 +1,512 @@
+
+	ts_scanner_foreach(&iterator)
+	{
+		bool isnull;
+		TupleInfo *ti_cagg = ts_scan_iterator_tuple_info(&iterator);
+		Datum cagg_hyper_id =
+			slot_getattr(ti_cagg->slot, Anum_continuous_agg_mat_hypertable_id, &isnull);
+		Assert(!isnull);
+		cagg_ids = lappend_int(cagg_ids, DatumGetInt32(cagg_hyper_id));
+	}
+
+	ts_scan_iterator_close(&iterator);
+
+	return cagg_ids;
+}
+
+static void
+hypertable_invalidation_scan_init(ScanIterator *iterator, int32 hyper_id, LOCKMODE lockmode)
+{
+	*iterator = ts_scan_iterator_create(CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG,
+										lockmode,
+										CurrentMemoryContext);
+	iterator->ctx.index = catalog_get_index(ts_catalog_get(),
+											CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG,
+											CONTINUOUS_AGGS_HYPERTABLE_INVALIDATION_LOG_IDX);
+	ts_scan_iterator_scan_key_init(
+		iterator,
+		Anum_continuous_aggs_hypertable_invalidation_log_idx_hypertable_id,
+		BTEqualStrategyNumber,
+		F_INT4EQ,
+		Int32GetDatum(hyper_id));
+}
+
+static HeapTuple
+create_invalidation_tup(const TupleDesc tupdesc, int32 cagg_hyper_id, int64 modtime, int64 start,
+						int64 end)
+{
+	Datum values[Natts_continuous_aggs_materialization_invalidation_log];
+	bool isnull[Natts_continuous_aggs_materialization_invalidation_log] = { false };
+
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_materialization_invalidation_log_materialization_id)] =
+		Int32GetDatum(cagg_hyper_id);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_materialization_invalidation_log_modification_time)] =
+		Int64GetDatum(modtime);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_materialization_invalidation_log_lowest_modified_value)] =
+		Int64GetDatum(start);
+	values[AttrNumberGetAttrOffset(
+		Anum_continuous_aggs_materialization_invalidation_log_greatest_modified_value)] =
+		Int64GetDatum(end);
+
+	return heap_form_tuple(tupdesc, values, isnull);
+}
+
+/*
+ * This format is common for tuples in both the hypertable invalidation log
+ * and the cagg invalidation log.
+ */
+typedef struct InvalidationEntry
+{
+	int32 id;
+	int64 modification_time;
+	int64 lowest_modified_value;
+	int64 greatest_modified_value;
+} InvalidationEntry;
+
+typedef enum InvalidationResult
+{
+	INVAL_NOMATCH,
+	INVAL_DELETE,
+	INVAL_CUT,
+} InvalidationResult;
+
+/*
+ * Try to cut an invalidation against the refresh window.
+ *
+ * If an invalidation entry overlaps with the refresh window, it needs
+ * additional processing: it is either cut, deleted, or left unmodified.
+ */
+static InvalidationResult
+cut_invalidation_along_refresh_window(const CaggInvalidationState *state,
+									  const InvalidationEntry *entry,
+									  const InternalTimeRange *refresh_window,
+									  const ItemPointer tid)
+{
+	int32 cagg_hyper_id = state->cagg.data.mat_hypertable_id;
+	TupleDesc tupdesc = RelationGetDescr(state->cagg_log_rel);
+	InvalidationResult result = INVAL_NOMATCH;
+	HeapTuple tup1 = NULL;
```

Comment:
They don't represent these ranges, strictly. The way it is used is: if one tuple is set, do only an update, if two tuples are set, then do one update and one insert (doesn't matter which one is update vs insert).

I can try to rework this to make the naming more clear here, but it will just move the problem around to another place.